reference-implementation/__tests__/composition.js

```diff
@@ -0,0 +1,367 @@
+          }
+        }
+      },
+      {
+        imports: {
+          'std:built-in': 'https://built-in-enhancement-3/'
+        },
+        scopes: {
+          'https://built-in-enhancement-3/': {
+            'std:built-in': 'std:built-in'
+          }
+        }
+      }
+    ])).toStrictEqual({
+      imports: { 'std:built-in': ['https://built-in-enhancement-3/'] },
+      scopes: {
+        'https://built-in-enhancement-1/': { 'std:built-in': ['std:built-in'] },
+        'https://built-in-enhancement-2/': { 'std:built-in': ['https://built-in-enhancement-1/'] },
+        'https://built-in-enhancement-3/': { 'std:built-in': ['https://built-in-enhancement-2/'] }
+      }
+    });
+  });
+
+  it('should compose "nested" scopes', () => {
+    expect(composeMaps([
+      {
+        imports: { 'https://a/': 'https://b/' },
+        scopes: {
+          'https://example.com/x/y/': { 'https://c/': 'https://d/' },
+          'https://example.com/x/y/z': { 'https://e/': 'https://f/' }
+        }
+      },
+      {
+        imports: { 'https://m/': 'https://n/' },
+        scopes: {
+          'https://example.com/x/y/z': {
+            'https://g/': 'https://a/',
+            'https://h/': 'https://c/',
+            'https://i/': 'https://e/'
+          }
+        }
+      }
+    ])).toStrictEqual({
+      imports: {
+        'https://a/': ['https://b/'],
+        'https://m/': ['https://n/']
+      },
+      scopes: {
+        'https://example.com/x/y/': { 'https://c/': ['https://d/'] },
+        'https://example.com/x/y/z': {
+          'https://e/': ['https://f/'],
+          'https://g/': ['https://b/'],
+          'https://h/': ['https://d/'],
+          'https://i/': ['https://f/']
+        }
+      }
+    });
+  });
+
+  it('should not clobber earlier more-specific scopes with later less-specific scopes', () => {
+    expect(composeMaps([
+      {
+        imports: {},
+        scopes: {
+          'https://example.com/x/y/': { 'https://a/': 'https://b/' },
+          'https://example.com/x/y/z': { 'https://c/': 'https://d/' }
+        }
+      },
+      {
+        imports: {
+          'https://a/': 'https://e/'
+        },
+        scopes: {
+          'https://example.com/x/': {
+            'https://c/': 'https://f/'
+          }
+        }
+      }
+    ])).toStrictEqual({
+      imports: {
+        'https://a/': ['https://e/']
+      },
+      scopes: {
+        'https://example.com/x/': { 'https://c/': ['https://f/'] },
+        'https://example.com/x/y/': { 'https://a/': ['https://b/'] },
+        'https://example.com/x/y/z': { 'https://c/': ['https://d/'] }
+      }
+    });
+  });
+
+  it('should perform package-prefix-relative composition', () => {
+    expect(composeMaps([
+      {
+        imports: {
+          'moment/': '/node_modules/moment/src/'
+        },
+        scopes: {}
+      },
+      {
+        imports: {
+          'utils/': 'moment/lib/utils/',
+          'is-date': 'moment/lib/utils/is-date.js'
```

Comment:
Can you test that this *doesn't* work within a single map? For example an entry

```js
'foo': 'utils/foo.js'
```

should result in... just a bare specifier in the output, I think?

Reply:
> should result in... just a bare specifier in the output, I think?

Yes, except that [this](https://github.com/WICG/import-maps/pull/167#discussion_r314064807) will change "the output is just a bare specifier" to the equivalent "the output is an empty array" (and a warning).